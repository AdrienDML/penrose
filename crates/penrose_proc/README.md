## proc-macros for the Penrose tiling window manager library

Most (if not all) of the proc-macros in this crate are consumed internally in
the main [penrose][1] crate and are not intended for use outside of that
context. Please see the main penrose crate for further details.

  [1]: https://crates.io/crates/penrose
