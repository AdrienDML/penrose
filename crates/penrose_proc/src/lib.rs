//! Proc macros for use in the main Penrose crate
use proc_macro::TokenStream;

mod stub;
mod validate_bindings;

use stub::stubbed_companion_trait_inner;
use validate_bindings::validate_user_bindings_inner;

/// This is an internal macro that is used as part of `gen_keybindings` to validate user provided
/// key bindings at compile time using xmodmap.
///
/// It is not intended for use outside of that context and may be modified and updated without
/// announcing breaking API changes.
///
/// ```no_run
/// # use penrose_proc::validate_user_bindings;
/// validate_user_bindings!(
///     ( "M-a", "M-b" )
///     (
///         ( ( "M-{}", "M-S-{}" ) ( "1", "2", "3" ) )
///     )
/// );
/// ```
#[proc_macro]
pub fn validate_user_bindings(input: TokenStream) -> TokenStream {
    validate_user_bindings_inner(input)
}

#[proc_macro_attribute]
pub fn stubbed_companion_trait(args: TokenStream, input: TokenStream) -> TokenStream {
    stubbed_companion_trait_inner(args, input)
}
