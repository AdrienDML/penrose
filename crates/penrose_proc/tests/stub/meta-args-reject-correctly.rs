// Invalid meta args should get rejected
use penrose_proc::stubbed_companion_trait;

#[allow(dead_code)]
#[derive(Eq, PartialEq, Debug)]
enum Scale {
    Major,
    Minor,
    Blues,
}

#[stubbed_companion_trait(bork = "CustomPrefix")]
trait Bar {
    #[stub("F#")]
    fn musical_note(&self, x: u32) -> &str;

    #[stub(Scale::Blues)]
    fn musical_scale(&self, f: f64) -> Scale;
}

fn main() {}
