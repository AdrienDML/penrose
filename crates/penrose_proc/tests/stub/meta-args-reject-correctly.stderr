error: Expected #[stubbed_companion_trait] or #[stubbed_companion_trait(prefix = "Foo")]
  --> $DIR/meta-args-reject-correctly.rs:13:27
   |
13 | #[stubbed_companion_trait(bork = "CustomPrefix")]
   |                           ^^^^^^^^^^^^^^^^^^^^^
